File: site/examples/rules.py
"""Rules the model can't break: a learned Score ranks actions, a written mask decides
which actions exist right now, and Decide never selects a masked one.

Runs offline on CPU:  pip install 'tensorcode[vec]'  then  python rules.py
"""
import re
import torch
from tensorcode import trace, training
from tensorcode.ops.vec import CandidateSet, Decide, Latent, Score
from tensorcode.ops.vec.encode import VocabularyEncoder

ACTIONS = {
    "restart": "restart the service process",
    "rollback": "roll back the latest deploy",
    "scale": "scale out more instances",
    "page": "page the on call engineer",
}
POSTMORTEMS = [
    ("errors started right after the deploy", "rollback"),
    ("the new release throws exceptions", "rollback"),
    ("memory keeps growing until the process dies", "restart"),
    ("the service hung and stopped responding", "restart"),
    ("traffic spike and requests are queueing", "scale"),
    ("cpu saturated during the sale", "scale"),
    ("the database disk is almost full", "page"),
    ("strange alerts from the payment provider", "page"),
]


def allowed(*, peak_hours, minutes_since_deploy, budget_left):
    return {
        "restart": not peak_hours,              # never at peak
        "rollback": minutes_since_deploy < 60,  # recent deploys only
        "scale": budget_left > 0,               # it costs money
        "page": True,                           # a person, always
    }


torch.manual_seed(5)
texts = [t for t, _ in POSTMORTEMS] + list(ACTIONS.values())
words = sorted({w for t in texts for w in re.findall(r"\w+|[^\w\s]", t)})
space = {"name": "operations text", "dimensions": 24}
encode = VocabularyEncoder({"vocabulary": words, "dimensions": 24, "output_space": space})
score = Score({"architecture": "mlp", "hidden_dimensions": [24], "query_space": space,
               "candidate_space": space, "meaning": "learned fit of an action to an incident"})
decide = Decide()


def options(incident, rules=None):
    actions = encode(list(ACTIONS.values()))
    mask = None if rules is None else torch.tensor([rules[a] for a in ACTIONS])
    return CandidateSet(encode(incident), Latent(actions.tensor, actions.space, mask=mask), list(ACTIONS))


sessions = []
for incident, action in POSTMORTEMS:
    with trace() as t:
        scores = score(options(incident))
    t.supervise(scores, list(ACTIONS).index(action), loss="choice", source="postmortem review")
    sessions.append(t)


def choice(scores, target):
    return torch.nn.functional.cross_entropy(scores.values.unsqueeze(0), torch.tensor([target]))


trainer = training.Trainer.from_ops({"encode": encode, "score": score}, losses={"choice": choice},
                                    optimizer=lambda p: torch.optim.Adam(p, lr=0.01))
losses = trainer.fit(sessions, epochs=25)
print(f"loss {losses[0]:.4f} -> {losses[-1]:.4f}")

incident = "errors after the deploy an hour ago"
with torch.no_grad():
    for minutes in (20, 90):
        rules = allowed(peak_hours=True, minutes_since_deploy=minutes, budget_left=100)
        scores = score(options(incident, rules))
        values = ", ".join(f"{a} {v:.2f}" for a, v in zip(ACTIONS, scores.values.tolist()))
        print(f"deploy {minutes:>2} min ago: {decide(scores).identity:8} scores: {values}")

File: site/assets/programs/rules.js
// site/examples/rules.py, run by TensorCode for TypeScript.
import * as tc from '../lib/tensorcode.js';

export const ACTIONS = {
  restart: 'restart the service process',
  rollback: 'roll back the latest deploy',
  scale: 'scale out more instances',
  page: 'page the on call engineer',
};
export const IDS = Object.keys(ACTIONS);
export const POSTMORTEMS = [
  ['errors started right after the deploy', 'rollback'],
  ['the new release throws exceptions', 'rollback'],
  ['memory keeps growing until the process dies', 'restart'],
  ['the service hung and stopped responding', 'restart'],
  ['traffic spike and requests are queueing', 'scale'],
  ['cpu saturated during the sale', 'scale'],
  ['the database disk is almost full', 'page'],
  ['strange alerts from the payment provider', 'page'],
];

export function allowed({ peakHours, minutesSinceDeploy, budgetLeft }) {
  return {
    restart: !peakHours, // never at peak
    rollback: minutesSinceDeploy < 60, // recent deploys only
    scale: budgetLeft > 0, // it costs money
    page: true, // a person, always
  };
}

const tokenize = (text) => text.toLowerCase().match(/\w+|[^\w\s]/g) ?? [];

export function build() {
  tc.manualSeed(5);
  const texts = [...POSTMORTEMS.map(([t]) => t), ...Object.values(ACTIONS)];
  const words = [...new Set(texts.flatMap(tokenize))].sort();
  const space = { name: 'operations text', dimensions: 24 };
  const encode = new tc.VocabularyEncoder({ vocabulary: words, dimensions: 24, output_space: space });
  const score = new tc.Score({
    architecture: 'mlp', hidden_dimensions: [24], query_space: space, candidate_space: space,
    meaning: 'learned fit of an action to an incident',
  });
  const decide = new tc.Decide();
  const options = (incident, rules = null) => {
    const actions = encode.call(Object.values(ACTIONS));
    const mask = rules === null ? null : tc.tensor(IDS.map((id) => rules[id]), { dtype: 'bool' });
    return new tc.CandidateSet(encode.call(incident), new tc.Latent(actions.tensor, actions.space, { mask }), IDS);
  };
  const sessions = POSTMORTEMS.map(([incident, action]) => {
    const t = tc.trace();
    const scores = t.run(() => score.call(options(incident)));
    t.supervise(scores, IDS.indexOf(action), { loss: 'choice', source: 'postmortem review' });
    return t;
  });
  const choice = (scores, target) => tc.crossEntropy(scores.values.unsqueeze(0), tc.tensor([target], { dtype: 'int64' }));
  const trainer = tc.Trainer.fromOps({ encode, score }, { losses: { choice }, optimizer: (p) => new tc.Adam(p, { lr: 0.01 }) });
  return {
    words: new Set(words),
    epoch: () => trainer.fit(sessions, { epochs: 1 }),
    /** Learned scores for every action (no mask), and the decision under `rules`. */
    run: (incident, rules) => tc.noGrad(() => {
      const learned = score.call(options(incident)).values.tolist();
      const scores = score.call(options(incident, rules));
      return { learned, masked: scores.values.tolist(), choice: decide.call(scores).identity };
    }),
  };
}

export function main(print) {
  const model = build();
  const losses = [];
  for (let epoch = 0; epoch < 25; epoch += 1) losses.push(...model.epoch());
  print(`loss ${losses[0].toFixed(4)} -> ${losses.at(-1).toFixed(4)}`);
  const incident = 'errors after the deploy an hour ago';
  for (const minutes of [20, 90]) {
    const rules = allowed({ peakHours: true, minutesSinceDeploy: minutes, budgetLeft: 100 });
    const { masked, choice } = model.run(incident, rules);
    const values = IDS.map((id, i) => `${id} ${masked[i].toFixed(2)}`).join(', ');
    print(`deploy ${String(minutes).padStart(2)} min ago: ${choice.padEnd(8)} scores: ${values}`);
  }
}
